src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -224,12 +226,20 @@ internal CompilationUnitSyntax ParseCompilationUnitCore()
             SyntaxListBuilder modifiers)
         {
             _recursionDepth++;
-            StackGuard.EnsureSufficientExecutionStack(_recursionDepth);
-            var result = ParseNamespaceDeclarationCore(attributeLists, modifiers);
+            var result = StackGuard.TryEnsureSufficientExecutionStack(_recursionDepth, _throwOnInsufficientExecutionStack) ?
+                this.ParseNamespaceDeclarationCore(attributeLists, modifiers) :
+                this.ParseNamespaceDeclarationCore_NewExecutionStack(attributeLists, modifiers);
             _recursionDepth--;
             return result;
         }
 
+        private NamespaceDeclarationSyntax ParseNamespaceDeclarationCore_NewExecutionStack(
```

Comment:
nit: if used in a single method, consider inlining as a local function, or completely inlining.

Reply:
I avoided adding local functions or inlining the calls to avoid changing the size of the stack frames of the existing callers, so that scenarios that could be parsed with on a single stack today can continue to be.